### PR TITLE
Implement BITFIELD command in test executor

### PR DESCRIPTION
**Pull request by mvelimir**

### CODE REVIEW

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 1]
```diff
@@ -2943,6 +2943,236 @@ private[redis] final class TestExecutor private (
...
+        }
+
+        def decodeOffset(s: String): Option[Int] = s match {
+          case offsetRegex("#", number) => toIntOption(number).filter(_ >= 0).map(_ * 8)
+          case offsetRegex("", number)  => toIntOption(number).filter(_ >= 0)
+          case _                        => None
+        }
+
+        def parseSignedLong(string: String): Option[Long] =
+          Some(string).filter(_.length <= 64).flatMap { string =>
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:2969]:
```suggestion
          Option(string).filter(_.length <= 64).flatMap { string =>
```

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 2]
```diff
@@ -2943,6 +2943,236 @@ private[redis] final class TestExecutor private (
...
+        import BitFieldCommand.BitFieldOverflow
+
+        def decodeBitFieldType(s: String): Option[BitFieldType] = s match {
+          case unsignedIntRegex(size) => toIntOption(size).filter(_ <= 64).map(size => BitFieldType.UnsignedInt(size))
+          case signedIntRegex(size)   => toIntOption(size).filter(_ <= 63).map(size => BitFieldType.SignedInt(size))
+          case _                      => None
+        }
+
+        def decodeOffset(s: String): Option[Int] = s match {
+          case offsetRegex("#", number) => toIntOption(number).filter(_ >= 0).map(_ * 8)
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:2963]: `filter` and `map` can be replaced with `collect`.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 3]
```diff
@@ -2943,6 +2943,236 @@ private[redis] final class TestExecutor private (
...
+
+        val unsignedIntRegex = "u(\\d+)".r
+        val signedIntRegex   = "i(\\d+)".r
+        val offsetRegex      = "([#]?)(\\d+)".r
+
+        import BitFieldCommand.BitFieldOverflow
+
+        def decodeBitFieldType(s: String): Option[BitFieldType] = s match {
+          case unsignedIntRegex(size) => toIntOption(size).filter(_ <= 64).map(size => BitFieldType.UnsignedInt(size))
+          case signedIntRegex(size)   => toIntOption(size).filter(_ <= 63).map(size => BitFieldType.SignedInt(size))
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:2958]: `filter` and `map` can be replaced with `collect`.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 4]
```diff
@@ -2943,6 +2943,236 @@ private[redis] final class TestExecutor private (
...
+          case BitFieldOverflow.Wrap =>
+            Some(value).map { value =>
+              bitFieldType match {
+                case BitFieldType.UnsignedInt(size) =>
+                  value % Math.pow(2, size.toDouble).toLong
+                case BitFieldType.SignedInt(size) =>
+                  value % (Math.pow(2, (size - 1).toDouble).toLong - 1L) - Math.pow(2, (size - 1).toDouble).toLong
+              }
+            }
+        }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3018]: This bit of code is quite difficult to read and involves a lot of repetition.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 5]
```diff
@@ -2943,6 +2943,236 @@ private[redis] final class TestExecutor private (
...
+                                    }.getOrElse((string, respChunk :+ RespValue.NullBulkString))
+                                  }
+                                }
+                            }
+                          }
+              } yield newResp
+
+              respOption match {
+                case Some(resp) => runCommands(tail, overflow, resp)
+                case None       => None
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3143]: This is a `map`.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 6]
```diff
@@ -2943,6 +2943,236 @@ private[redis] final class TestExecutor private (
...
+        orMissingParameter(keyOption) { key =>
+          orWrongType(isString(key))(
+            strings.getOrElse(key, "").flatMap { string =>
+              runCommands(stringInput.tail.toList, BitFieldOverflow.Wrap, STM.succeed((string, Chunk.empty))).map {
+                resp =>
+                  resp.flatMap { case (string, respChunk) =>
+                    strings.put(key, string) *> STM
+                      .succeed(if (respChunk.isEmpty) RespValue.NullArray else RespValue.Array(respChunk))
+                  }.orElseSucceed(Replies.Error)
+              }.getOrElse(STM.succeed(Replies.Error))
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3171]: I have no clue what's going on here :)

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 7]
```diff
@@ -2943,6 +2943,236 @@ private[redis] final class TestExecutor private (
...
+          val substring       = newString.slice(offset / 8, (offset + size - 1) / 8 + 1)
+          val binarySubstring = substring.foldLeft("")((a, b) => a + f"${b.toInt.toBinaryString}%8s".replace(' ', '0'))
+          val binaryValue     = value.toBinaryString
+          val newBinarySubstring =
+            binarySubstring.slice(0, offset % 8) + s"%${size}s".format(binaryValue).replace(' ', '0') + binarySubstring
+              .slice(offset % 8 + size, binarySubstring.length)
+          val newSubstring = newBinarySubstring.grouped(8).toList.map(x => Integer.parseInt(x, 2).toChar).mkString
+
+          newString.slice(0, offset / 8) + newSubstring + newString.slice((offset + size - 1) / 8 + 1, newString.length)
+        }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3039]: Ditto.

**Review by mijicd — CHANGES_REQUESTED** [review]

File: redis/src/main/scala/zio/redis/TestExecutor.scala (hunk 5 above)
Comment (mvelimir) [reply redis/src/main/scala/zio/redis/TestExecutor.scala:3143]: I wanted to make the function tail-recursive.

**Comment by mijicd** [comment]
Deprecated by #747.

**Pull request closed without merging**